build.sbt

```diff
@@ -56,8 +56,14 @@ mimaBinaryIssueFilters ++= Seq(
   ProblemFilters.exclude[ReversedMissingMethodProblem](
     "io.github.davidgregory084.ScalacOptions.privatePartialUnification"
   ),
+  ProblemFilters.exclude[ReversedMissingMethodProblem](
+    "io.github.davidgregory084.ScalacOptions.async"
+  ),
   ProblemFilters.exclude[ReversedMissingMethodProblem](
     "io.github.davidgregory084.ScalacOptions.io$github$davidgregory084$ScalacOptions$_setter_$privatePartialUnification_="
+  ),
+  ProblemFilters.exclude[ReversedMissingMethodProblem](
+    "io.github.davidgregory084.ScalacOptions.io$github$davidgregory084$ScalacOptions$_setter_$async_="
   )
```

Comment:
Hmm, none of these should be filtered since none of these are binary-compatible actually 😕

Reply:
Should we guarantee forward binary compatibility? IMO I don't think so

Reply:
I should clarify - for this trait specifically, it is not part of `autoImport` and it's not intended that users will extend it. I don't think that adding new methods to it should cause anybody a problem.

Reply:
The issue here is backwards binary compatibility actually. In any case, if this trait is not meant to be extended by user code it should be sealed and possibly even private.

Reply:
Hmm can you explain that a bit more? My understanding is that forward compatibility means that an older version of sbt-tpolecat could be used where a newer version is expected. Adding a new method does not break that constraint unless I misunderstand how it works (which I probably do). You are right that this trait should be locked down more and I think it makes sense to do that as part of the breaking changes saved up for the next release.

Reply:
> Adding a new method does not break that constraint unless I misunderstand how it works (which I probably do).

Right, exactly, _none_ of us understand how it works, which is exactly why we run MiMa :) you are right, that at a source-level the change here looks perfectly fine: adding a new method _with a default implementation_ should be no problem. Note that adding a new abstract method would be though.

However, unfortunately the Scala compiler has encoded this change at the bytecode-level using abstract methods, which is exactly why MiMa is upset.

Changing all the affected `val`'s to `def`s should allow you to remove the filters.

Reply:
> However, unfortunately the Scala compiler has encoded this change at the bytecode-level using abstract methods, which is exactly why MiMa is upset.
That is still a forward-compatibility problem though right? Rather than a backwards compatible one? I'm trying to understand what a user would have to do in order to hurt themselves with this.
It would certainly be a very atypical use of this plugin I think? Would it not require sbt to evict the sbt-tpolecat version downwards in case of a conflict, and that the user had extended the trait themselves for some reason? Then they could see a `NoSuchMethodError` when their build attempted to use one of the newer methods of `ScalacOptions`.

Reply:
No, it's backwards compatibility unfortunately. All it takes is for me to implement this `trait` in sbt-typelevel and publish, for you to make a change like you've done here and publish, and for a user to mix the the two plugins in the same build. My sbt-typelevel implementation will get into trouble because it will be missing implementations of these new abstract methods that have been added to the `trait`.

Reply:
Hmm ok, I really struggled to find any documentation about this - in the end I could only find two snippets that referenced `ReversedMissingMethodProblem` at all and one general piece of advice:

* https://github.com/lightbend/mima/issues/205 which says that it's not a problem as of 2.12
* The Binary compatibility examples docs: https://github.com/jatcwang/binary-compatibility-guide#dont-adding-methods-with-default-implementation-to-traits-211-or-before= which says this only affects 2.11 or lower
* The Akka docs https://doc.akka.io/docs/akka-http/10.0/contributing.html#binary-compatibility which say that this one can probably be ignored when `it is adding API to classes / traits which are only meant for extension by Akka itself, i.e. should not be extended by end-users`

In the end, since we only support sbt 1.x which uses 2.12, and none of the scala-steward builds that I could see were broken due to bincompat I presumed it was not causing users any problems, but the critical part that I was missing was that I had not really locked down that trait in any way.

I guess I can put adding example scenarios to the MiMa docs for each possible problem to the queue of stuff that I can work on. 😛

Reply:
> which says that it's not a problem as of 2.12

Unfortunately, there is some subtlety there. It used to be that adding _any_ `def` to a `trait` (even with a default implementation) was binary-breaking. As of 2.12 that is no longer the case. _However_, adding a `val` as you've done here is still breaking IIUC.

> and none of the scala-steward builds that I could see were broken due to bincompat I presumed it was not causing users any problems

Right, because every build recompiles the source from scratch :) the specific conditions to observe this problem are that sbt-tpolecat is depended on by another sbt plugin, that sbt plugin extends this trait, and a user includes both plugins in their build.

Binary compatibility never causes problems directly for user code, because they are always recompiling from scratch. It causes problems for other libraries they are including in their project however, since they have been compiled against older versions of your library.